spec/p2p/v0.34/pex-protocol.md

```diff
@@ -0,0 +1,240 @@
+
+While all nodes, with few exceptions, participate on the PEX protocol,
+a subset of nodes, configured as [seed nodes](#seed-nodes) have a particular
+role in the protocol.
+They crawl the network, connecting to random peers, in order to learn as many
+peer addresses as possible to provide to other nodes.
+
+## Requesting Addresses
+
+A node requests peer addresses by sending a `PexRequest` message to a peer.
+
+For regular nodes, not operating in seed mode, a PEX request is sent when
+the node *needs* peers addresses, a condition checked:
+
+1. When an *outbound* peer is added, causing the node to request addresses from
+   the new peer
+2. Periodically, by the `ensurePeersRoutine`, causing the node to request peer
+   addresses to a randomly selected peer
+
+A node needs more peer addresses when its addresses book has
+[less than 1000 records](./addressbook.md#need-for-addresses).
+It is thus reasonable to assume that the common case is that a peer needs more
+peer addresses, so that PEX requests are sent whenever the above two situations happen.
+
+A PEX request is sent when a new *outbound* peer is added.
+The same does not happen with new inbound peers
+because outbound peers, that the node has dialed, are considered more trustworthy
+than inbound peers, that the node has accepted.
+Moreover, when a node is short of peer addresses, it dials the configured seed nodes;
+since they are added as outbound peers, the node can immediately request peer addresses.
+
+The `ensurePeersRoutine` periodically checks, by default every 30 seconds (`ensurePeersPeriod`),
+whether the node has enough outbound peers.
+If it does not have, the node tries dialing some peer addresses stored in the address book.
+As part of this procedure, the node selects a peer at random,
+from the set of connected peers retrieved from the switch,
+and sends a PEX request to the selected peer.
+
+Sending a PEX request to a peer is implemented by the `RequestAddrs` method of
+the PEX reactor.
+
+### Responses
+
+After a PEX request is sent to a peer, the node expects to receive,
+as a response, a `PexAddrs` message from the peer.
+This message encodes a list of peer addresses that are
+[added to address book](./addressbook.md#adding-addresses),
+having the peer from which the PEX response was received as their source.
+
+Received PEX responses are handled by the `ReceiveAddrs` method of the PEX reactor.
+In the case of a PEX response received from a peer which is configured as
+a seed node, the PEX reactor attempts immediately to dial the provided peer
+addresses, as detailed [here](./peer_manager.md#fast-dialing).
+
+### Misbehavior
+
+Sending multiple PEX requests to a peer, before receiving a reply from it,
+is considered a misbehavior.
+To prevent it, the node maintains a `requestsSent` set of outstanding
+requests, indexed by destination peers.
+While a peer ID is present in the `requestsSent` set, the node does not send
+further PEX requests to that peer.
```

Comment:
Are there any provisions on the receiving side to mark the misbehaving peer as a bad peer?